components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/UserSignUpWSWorkflowExecutor.java

```diff
@@ -123,66 +121,59 @@ public WorkflowResponse execute(WorkflowDTO workflowDTO) throws WorkflowExceptio
-
-    		if (WorkflowStatus.APPROVED.equals(workflowDTO.getStatus())) {
-    			try {
-    				updateRolesOfUser(serverURL, adminUsername, adminPassword, tenantAwareUserName,
-    				                  SelfSignUpUtil.getRoleNames(signupConfig), tenantDomain);
-    			} catch (Exception e) {
-    				throw new WorkflowException("Error while assigning role to user", e);
-    			}
-    		} else {
-    			try {
-    				/* Remove created user */
-    				deleteUser(serverURL, adminUsername, adminPassword,
-    				           tenantAwareUserName);
-    			} catch (Exception e) {
-    				throw new WorkflowException("Error while deleting the user", e);
-    			}
-    		}
-    	} catch (APIManagementException e1) {
-    		throw new WorkflowException("Error while accessing signup configuration", e1);
-    	}
-        return new GeneralWorkflowResponse();
-	}
+    public WorkflowResponse complete(WorkflowDTO workflowDTO) throws WorkflowException {
+
+        workflowDTO.setStatus(workflowDTO.getStatus());
+        workflowDTO.setUpdatedTime(System.currentTimeMillis());
+
+        if (log.isDebugEnabled()) {
+            log.debug("User Sign Up [Complete] Workflow Invoked. Workflow ID : " +
+                    workflowDTO.getExternalWorkflowReference() + "Workflow State : " +
+                    workflowDTO.getStatus());
+        }
+
+        super.complete(workflowDTO);
+
+        APIManagerConfiguration config = ServiceReferenceHolder.getInstance().getAPIManagerConfigurationService()
+                .getAPIManagerConfiguration();
+        String serverURL = config.getFirstProperty(APIConstants.AUTH_MANAGER_URL);
+
+        String tenantDomain = workflowDTO.getTenantDomain();
+        try {
 
+            UserRegistrationConfigDTO signupConfig = SelfSignUpUtil.getSignupConfiguration(tenantDomain);
+
+            String adminUsername = signupConfig.getAdminUserName();
+            String adminPassword = signupConfig.getAdminPassword();
+            if (serverURL == null || adminUsername == null || adminPassword == null) {
+                throw new WorkflowException("Required parameter missing to connect to the" + " authentication manager");
+            }
+
+            String tenantAwareUserName = MultitenantUtils.getTenantAwareUsername(workflowDTO.getWorkflowReference());
+
+            if (WorkflowStatus.APPROVED.equals(workflowDTO.getStatus())) {
+                try {
+                    updateRolesOfUser(serverURL, adminUsername, adminPassword, tenantAwareUserName,
+                            SelfSignUpUtil.getRoleNames(signupConfig), tenantDomain);
+                } catch (Exception e) {
+                    throw new WorkflowException("Error while assigning role to user", e);
+                }
+            } else {
+                try {
+                    /* Remove created user */
+                    deleteUser(serverURL, adminUsername, adminPassword, tenantAwareUserName);
+                } catch (Exception e) {
```

Comment:
I believe it is done that way to report the specific error message. Otherwise both failures in if and else statements will have the same error message.